### ISSUE TITLE
Proper Typing for Events

### ISSUE THREAD

**Issue by zackdotcomputer** [nextauthjs/next-auth#1639]
**Summary of proposed feature**
Since the arguments to each event in the `events` object are known, the typing should include that information.

**Purpose of proposed feature**
To making coding easier and prevent errors when using `events`

**Detail about proposed feature**
Currently the events object has the following typing:
```typescript
export type EventCallback = (message: any) => Promise<void>;
export type EventsOptions = Partial<Record<EventType, EventCallback>>;
```

I think it should have an actual shape defining the [event names supported](https://next-auth.js.org/configuration/events) and their message types. As an example:

```typescript
export type EventCallback<MessageType = any> = (message: MessageType) => Promise<void>;

type SignInMessage = {
  user: User;
  account: Record<string, unknown>;
  isNewUser?: boolean;
};

export type EventsOptions = Partial<{
  signIn: EventCallback<SignInMessage>;
  // And so on...
}>;
```

**Potential problems**
I don't foresee any issues with this. The only thing I'm unsure about is if your .d.ts files are autogenerated or manually written.

*Please indicate if you are willing and able to help implement the proposed feature.*
Of course, just let me know if your .d.ts files are autogenerated or manually written.

**Comment by zackdotcomputer** [comment]
Sorry - just saw your note on the npm package page that typing issues should be directed to DefinitelyTyped. Maybe the .d.ts files in this project should be removed?

Moving my issue over to them.

**Comment by balazsorban44** [comment]
The .d.ts types in this repo are not intended for use, that's why we don't advertise them. Currently, they are only used to improve local DX of the package. One day I might be comfortable with exporting them for everyone to use, but right now I don't feel confident about them, also they are not complete and I don't wish to create maintainer overhead by also starting to answer TS questions. I'm no expert in the topic.

**Comment by zackdotcomputer** [comment]
👍 I saw the note in the README about that after opening this issue - I created a [PR over here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52152) that you can review to update the typing for Events in the definitelytyped repo, @balazsorban44.

I understand not wanting to stop everything and do a deep rewrite - though on the consumer end (especially as Next itself continues to tighten its TS integrations) I find it a lot easier to use packages with solid typing files. I've been working with TS for ~5 years now, so I'll send over more PRs in that repo if I find further rough edges that could use refinement.

**Comment by balazsorban44** [comment]
https://github.com/nextauthjs/next-auth/issues/1649 Might be interesting to you then. :) We ask for a bit of patience, but we realized that the separation of types from the core is hard for those who maintain the types already. I really wished it would have been a more community effort, but it seems like that it is already only a few people who have maintained them. It will be easier to collect everything under the same repo for them. So if you want to open some PRs, it is very appreciated, but if you wait a maybe a few days, you might be able to open those against this repo... 😉

**Comment by zackdotcomputer** [comment]
Haha amazing.

Ok, I've subscribed to notifications on that PR. I'll move *my* PR over from DefinitelyTyped to this repo once that one lands.

**Comment by zackdotcomputer** [comment]
I guess I will reopen this issue, then, until that all is done.

**Comment by balazsorban44** [comment]
We will have a look at the migration today, together with @lluia. If you want to be more involved with providing more up-to-date types in the future, you can write me an e-mail, or at Twitter (`balazsorban44`) 😊

**Comment by balazsorban44** [comment]
@zackdotcomputer TS support has been out sine `3.15.0`, feel free to open a new PR with your suggestions!

Here is the relevant part: https://github.com/nextauthjs/next-auth/blob/065d9eb310cd873f2aa80b01bfb97d6d4a2bcceb/types/index.d.ts#L342-L356

**Comment by zackdotcomputer** [comment]
@balazsorban44 That's great! I've been pretty busy with a new puppy 🐶 but managed to sneak a PR in during one of her naps.
